pkg/controller/rolloutblock/rolloutblock_controller.go

```diff
@@ -0,0 +1,622 @@
+		go wait.Until(c.runRolloutBlockWorker, time.Second, stopCh)
+	}
+
+	glog.V(4).Info("Started RolloutBlock controller")
+
+	<-stopCh
+}
+
+func (c *Controller) runRolloutBlockWorker() {
+	for c.processNextRolloutBlockWorkItem() {
+	}
+}
+
+func (c *Controller) processNextRolloutBlockWorkItem() bool {
+	obj, shutdown := c.rolloutblockWorkqueue.Get()
+	if shutdown {
+		return false
+	}
+
+	defer c.rolloutblockWorkqueue.Done(obj)
+
+	var (
+		ok                  bool
+		rolloutBlockUpdater RolloutBlockUpdater
+	)
+
+	if rolloutBlockUpdater, ok = obj.(RolloutBlockUpdater); !ok {
+		c.rolloutblockWorkqueue.Forget(obj)
+		runtime.HandleError(fmt.Errorf("invalid object key (will retry: false): %#v", obj))
+		return true
+	}
+
+	shouldRetry := false
+	err := c.syncRolloutBlock(rolloutBlockUpdater)
+
+	if err != nil {
+		shouldRetry = shippererrors.ShouldRetry(err)
+		runtime.HandleError(fmt.Errorf("error syncing RolloutBlock %q (will retry: %t): %s", rolloutBlockUpdater.RolloutBlockKey, shouldRetry, err.Error()))
+	}
+
+	if shouldRetry {
+		if c.rolloutblockWorkqueue.NumRequeues(rolloutBlockUpdater) >= maxRetries {
+			// Drop this update out of the workqueue and thus reset its
+			// backoff. This limits the time a "broken" object can hog a worker.
+			glog.Warningf("Update %q for RolloutBlock has been retried too many times, dropping from the queue", rolloutBlockUpdater)
+			c.rolloutblockWorkqueue.Forget(rolloutBlockUpdater)
+
+			return true
+		}
+
+		c.rolloutblockWorkqueue.AddRateLimited(rolloutBlockUpdater)
+
+		return true
+	}
+
+	glog.V(4).Infof("Successfully synced RolloutBlock Updater %q", rolloutBlockUpdater)
+	c.rolloutblockWorkqueue.Forget(obj)
+
+	return true
+}
+
+func (c *Controller) syncRolloutBlock(rolloutBlockUpdater RolloutBlockUpdater) error {
+	key := rolloutBlockUpdater.RolloutBlockKey
+	updaterType := rolloutBlockUpdater.UpdaterType
+	if updaterType == NewRolloutBlockObject {
+		return c.syncNewRolloutBlockObject(key)
+	}
+
+	overridingKey := rolloutBlockUpdater.OverridingObjectKey
+	if len(key) == 0 || len(overridingKey) == 0 {
+		return nil
+	}
+
+	var err error = nil
+	isDeletedObject := rolloutBlockUpdater.IsDeletedObject
+
+	if isDeletedObject {
+		switch updaterType {
+		case OverridingRelease:
+			err = c.removeReleaseFromRolloutBlockStatus(overridingKey, key)
+		case OverridingApplication:
+			err = c.removeAppFromRolloutBlockStatus(overridingKey, key)
+		}
+	} else {
+		switch updaterType {
+		case OverridingRelease:
+			err = c.addReleaseToRolloutBlockStatus(overridingKey, key)
+		case OverridingApplication:
+			err = c.addApplicationToRolloutBlockStatus(overridingKey, key)
+		}
+	}
+
+	return err
+}
+
+func (c *Controller) syncNewRolloutBlockObject(key string) error {
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	rolloutBlock, err := c.rolloutBlockLister.RolloutBlocks(ns).Get(name)
+	if err != nil {
+		return err
+	}
+
+	if ns == shipper.ShipperNamespace {
```

Comment:
You can preserve a corresponding lister function, like:
```go
var listerFunc <here comes it's type>
switch ns {
case shipper.ShipperNamespace:
    listerFunc = c.applicationLister.List
default:
    listerFunc := c.applicationLister.Applications(ns).List
}
This will allow you to drop this if-else statement completely
```